### ISSUE TITLE
Question about the MLP layers.

### ISSUE THREAD

**Issue by guaguabujianle** [gregory-kyro/HAC-Net#3]
Dear Gregory W.Kyro,
Thank you for sharing the code. It is really helpful. However, I have some question about the MLP layer. To my understanding, the 3D CNN already have a MLP layer to compress the learned features to a scalar value corresponding to pKa value. Why do you discard that MLP in 3D-CNN after training and re-train a new one?

Best,
Yang

**Comment by guaguabujianle** [comment]
Dear Gregory W.Kyro,
Here is another question. In your processed data, do you retain water molecules? I've observed that during the preprocessing step, these molecules aren't removed. As I know, a lot of other methods would try to remove water molecules during preprocessing. Is that removing water molecules causing negative impact on the model performance? I am looking forward to your reply. Thank you very much.

Best,
Yang

**Comment by gregory-kyro** [comment]
Hi Yang,

**_Answer to question 1:_** 
After passed through the convolutional layers in the 3D-CNN, the data are flattened into a vector of length 2048 and passed to a fully connected layer of size 2048 × 100 with ReLU activation and batch normalization and then to a final fully connected layer of size 100 × 1. We do not discard the fully connected layers -- we train this 3D-CNN in its entirety. However, after the initial 3D-CNN training is complete, we extract the flattened features of size 2048 and use them to train a pair of fully connected layers identical to those in the 3D-CNN architecture. We do this because the fully connected layers account for only 1.9% of the total parameters in the 3D-CNN, causing the initial learning to be driven primarily by the parameters for the convolutional layers. We find that this protocol notably improves performance. Please let me know if this explanation is clear.

**_Answer to question 2:_** 
We decided to keep the water molecules since they are often important for the binding interaction occurring between the ligand and the pocket of the protein.

Please let me know if you have any other questions!

Best,
Gregory

**Comment by guaguabujianle** [comment]
Hi, Gregory. Thank you for quick reply. I have tried to train the CNN and GNN models. The GNN model works fine. However, I have some problems in training the CNN model. Specifically, the validation loss/performance is unstable. Sometime it will get a very high validation loss or very bad performance during training, is that normal? Below is the training logger. At 8-th epoch, the model get a very high loss.
[1/100-1/18] training loss: 136.917
[1/100-2/18] training loss: 80.394
[1/100-3/18] training loss: 240.886
[1/100-4/18] training loss: 71.888
[1/100-5/18] training loss: 60.835
[1/100-6/18] training loss: 70.968
[1/100-7/18] training loss: 94.933
[1/100-8/18] training loss: 37.948
[1/100-9/18] training loss: 22.469
[1/100-10/18] training loss: 27.311
[1/100-11/18] training loss: 23.691
[1/100-12/18] training loss: 23.471
[1/100-13/18] training loss: 18.507
[1/100-14/18] training loss: 13.076
[1/100-15/18] training loss: 19.798
[1/100-16/18] training loss: 20.422
[1/100-17/18] training loss: 15.087
[1/100-18/18] training loss: 15.507
[1/100-1/18] validation loss: 188.521
[1/100-2/18] validation loss: 94.232
[1/100-3/18] validation loss: 100.811
[1/100-4/18] validation loss: 108.904
[1/100] validation results-- pearsonr: 0.215, spearmanr: 0.239, rmse: 11.096, mae: 8.438, r2: -26.244
best checkpoint saved: content/cnn_best_parameters.pt
checkpoint saved: content/cnn_full_parameters.pt
[2/100-1/18] training loss: 13.194
[2/100-2/18] training loss: 15.606
[2/100-3/18] training loss: 13.400
[2/100-4/18] training loss: 12.136
[2/100-5/18] training loss: 8.601
[2/100-6/18] training loss: 12.230
[2/100-7/18] training loss: 11.725
[2/100-8/18] training loss: 11.637
[2/100-9/18] training loss: 12.587
[2/100-10/18] training loss: 7.631
[2/100-11/18] training loss: 7.400
[2/100-12/18] training loss: 8.864
[2/100-13/18] training loss: 5.594
[2/100-14/18] training loss: 6.649
[2/100-15/18] training loss: 7.716
[2/100-16/18] training loss: 6.155
[2/100-17/18] training loss: 5.265
[2/100-18/18] training loss: 4.258
[2/100-1/18] validation loss: 682.635
[2/100-2/18] validation loss: 677.477
[2/100-3/18] validation loss: 761.270
[2/100-4/18] validation loss: 634.746
[2/100] validation results-- pearsonr: -0.422, spearmanr: -0.458, rmse: 26.249, mae: 23.312, r2: -151.472
checkpoint saved: content/cnn_full_parameters.pt
[3/100-1/18] training loss: 5.751
[3/100-2/18] training loss: 7.306
[3/100-3/18] training loss: 5.607
[3/100-4/18] training loss: 5.798
[3/100-5/18] training loss: 4.984
[3/100-6/18] training loss: 7.070
[3/100-7/18] training loss: 3.642
[3/100-8/18] training loss: 3.766
[3/100-9/18] training loss: 5.791
[3/100-10/18] training loss: 6.819
[3/100-11/18] training loss: 5.708
[3/100-12/18] training loss: 3.682
[3/100-13/18] training loss: 6.633
[3/100-14/18] training loss: 6.162
[3/100-15/18] training loss: 5.804
[3/100-16/18] training loss: 3.008
[3/100-17/18] training loss: 4.012
[3/100-18/18] training loss: 5.058
[3/100-1/18] validation loss: 54.975
[3/100-2/18] validation loss: 79.416
[3/100-3/18] validation loss: 131.273
[3/100-4/18] validation loss: 64.825
[3/100] validation results-- pearsonr: 0.073, spearmanr: 0.224, rmse: 9.090, mae: 5.692, r2: -17.283
checkpoint saved: content/cnn_full_parameters.pt
[4/100-1/18] training loss: 3.641
[4/100-2/18] training loss: 3.310
[4/100-3/18] training loss: 5.001
[4/100-4/18] training loss: 3.465
[4/100-5/18] training loss: 3.951
[4/100-6/18] training loss: 3.947
[4/100-7/18] training loss: 3.365
[4/100-8/18] training loss: 3.403
[4/100-9/18] training loss: 5.498
[4/100-10/18] training loss: 3.091
[4/100-11/18] training loss: 3.896
[4/100-12/18] training loss: 2.142
[4/100-13/18] training loss: 2.627
[4/100-14/18] training loss: 4.907
[4/100-15/18] training loss: 4.479
[4/100-16/18] training loss: 3.824
[4/100-17/18] training loss: 3.507
[4/100-18/18] training loss: 5.064
[4/100-1/18] validation loss: 9.199
[4/100-2/18] validation loss: 10.175
[4/100-3/18] validation loss: 14.220
[4/100-4/18] validation loss: 132.291
[4/100] validation results-- pearsonr: 0.163, spearmanr: 0.440, rmse: 6.440, mae: 3.020, r2: -8.177
best checkpoint saved: content/cnn_best_parameters.pt
checkpoint saved: content/cnn_full_parameters.pt
[5/100-1/18] training loss: 5.187
[5/100-2/18] training loss: 2.438
[5/100-3/18] training loss: 4.466
[5/100-4/18] training loss: 4.092
[5/100-5/18] training loss: 3.172
[5/100-6/18] training loss: 6.008
[5/100-7/18] training loss: 4.311
[5/100-8/18] training loss: 6.959
[5/100-9/18] training loss: 4.798
[5/100-10/18] training loss: 6.638
[5/100-11/18] training loss: 6.136
[5/100-12/18] training loss: 3.486
[5/100-13/18] training loss: 3.509
[5/100-14/18] training loss: 3.788
[5/100-15/18] training loss: 2.630
[5/100-16/18] training loss: 2.902
[5/100-17/18] training loss: 8.662
[5/100-18/18] training loss: 5.600
[5/100-1/18] validation loss: 4.325
[5/100-2/18] validation loss: 2.681
[5/100-3/18] validation loss: 4.829
[5/100-4/18] validation loss: 5.365
[5/100] validation results-- pearsonr: 0.487, spearmanr: 0.523, rmse: 2.074, mae: 1.622, r2: 0.049
best checkpoint saved: content/cnn_best_parameters.pt
checkpoint saved: content/cnn_full_parameters.pt
[6/100-1/18] training loss: 3.375
[6/100-2/18] training loss: 3.477
[6/100-3/18] training loss: 3.224
[6/100-4/18] training loss: 4.189
[6/100-5/18] training loss: 4.371
[6/100-6/18] training loss: 2.538
[6/100-7/18] training loss: 2.810
[6/100-8/18] training loss: 2.880
[6/100-9/18] training loss: 2.926
[6/100-10/18] training loss: 2.664
[6/100-11/18] training loss: 2.890
[6/100-12/18] training loss: 2.875
[6/100-13/18] training loss: 2.613
[6/100-14/18] training loss: 2.765
[6/100-15/18] training loss: 2.394
[6/100-16/18] training loss: 2.879
[6/100-17/18] training loss: 3.685
[6/100-18/18] training loss: 1.942
[6/100-1/18] validation loss: 4.019
[6/100-2/18] validation loss: 3.697
[6/100-3/18] validation loss: 5.076
[6/100-4/18] validation loss: 5.371
[6/100] validation results-- pearsonr: 0.519, spearmanr: 0.527, rmse: 2.131, mae: 1.720, r2: -0.005
best checkpoint saved: content/cnn_best_parameters.pt
checkpoint saved: content/cnn_full_parameters.pt
[7/100-1/18] training loss: 2.072
[7/100-2/18] training loss: 2.402
[7/100-3/18] training loss: 2.537
[7/100-4/18] training loss: 2.673
[7/100-5/18] training loss: 4.130
[7/100-6/18] training loss: 2.557
[7/100-7/18] training loss: 3.139
[7/100-8/18] training loss: 1.808
[7/100-9/18] training loss: 1.848
[7/100-10/18] training loss: 2.169
[7/100-11/18] training loss: 2.642
[7/100-12/18] training loss: 3.374
[7/100-13/18] training loss: 3.131
[7/100-14/18] training loss: 2.925
[7/100-15/18] training loss: 2.252
[7/100-16/18] training loss: 2.763
[7/100-17/18] training loss: 3.323
[7/100-18/18] training loss: 4.745
[7/100-1/18] validation loss: 8.051
[7/100-2/18] validation loss: 5.173
[7/100-3/18] validation loss: 6.513
[7/100-4/18] validation loss: 13.541
[7/100] validation results-- pearsonr: 0.452, spearmanr: 0.529, rmse: 2.884, mae: 2.177, r2: -0.841
checkpoint saved: content/cnn_full_parameters.pt
[8/100-1/18] training loss: 2.712
[8/100-2/18] training loss: 2.765
[8/100-3/18] training loss: 3.182
[8/100-4/18] training loss: 3.760
[8/100-5/18] training loss: 4.842
[8/100-6/18] training loss: 13.853
[8/100-7/18] training loss: 19.528
[8/100-8/18] training loss: 9.770
[8/100-9/18] training loss: 9.275
[8/100-10/18] training loss: 3.719
[8/100-11/18] training loss: 4.044
[8/100-12/18] training loss: 4.826
[8/100-13/18] training loss: 3.480
[8/100-14/18] training loss: 3.569
[8/100-15/18] training loss: 2.876
[8/100-16/18] training loss: 4.692
[8/100-17/18] training loss: 8.999
[8/100-18/18] training loss: 4.628
[8/100-1/18] validation loss: 292955.250
[8/100-2/18] validation loss: 70747.703
[8/100-3/18] validation loss: 149357520.000
[8/100-4/18] validation loss: 155339.109
[8/100] validation results-- pearsonr: 0.096, spearmanr: -0.286, rmse: 6121.204, mae: 625.511, r2: -8291339.061
checkpoint saved: content/cnn_full_parameters.pt

**Comment by gregory-kyro** [comment]
Hi Yang,

Yes, this is normal. I would recommend training the 3D-CNN for at least 50 epochs, preferably 100 epochs. We also provide all of our parameter files if that would make things easier for you. Please let me know if you have any more questions!

**Comment by guaguabujianle** [comment]
Thank you for you help.
